remote_attestation/rust/src/tests/message.rs

```diff
@@ -0,0 +1,222 @@
+//
+// Copyright 2021 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use crate::{
+    crypto::{
+        KEY_AGREEMENT_ALGORITHM_KEY_LENGTH, NONCE_LENGTH, SIGNATURE_LENGTH,
+        SIGNING_ALGORITHM_KEY_LENGTH,
+    },
+    message::{
+        deserialize_message, ClientHello, ClientIdentity, Deserializable, EncryptedData,
+        MessageWrapper, Serializable, ServerIdentity, MAXIMUM_MESSAGE_SIZE,
+        REPLAY_PROTECTION_ARRAY_LENGTH,
+    },
+};
+use anyhow::{anyhow, Context};
+use assert_matches::assert_matches;
+use quickcheck::{quickcheck, TestResult};
+
+const INVALID_MESSAGE_HEADER: u8 = 5;
+
+/// Creates a zero initialized array.
+fn default_array<T, const L: usize>() -> [T; L]
+where
+    T: std::marker::Copy + std::default::Default,
+{
+    [Default::default(); L]
+}
+
+/// Converts slices to arrays (expands with zeroes).
+fn to_array<T, const L: usize>(input: &[T]) -> anyhow::Result<[T; L]>
+where
+    T: std::marker::Copy + std::default::Default,
+{
+    if input.len() <= L {
+        let mut result: [T; L] = default_array();
+        result[..input.len()].copy_from_slice(&input[..input.len()]);
+        Ok(result)
+    } else {
+        Err(anyhow!(
+            "Maximum input length exceeded, maximum {}, found {}",
+            L,
+            input.len()
+        ))
+    }
+}
+
+fn test_serialize_template<M: Serializable + Deserializable + PartialEq>(
+    message: &M,
+) -> anyhow::Result<bool> {
+    let serialized_message = message.serialize().context("Couldn't serialize message")?;
+    let deserialized_message =
+        M::deserialize(&serialized_message).context("Couldn't deserialize message")?;
+    Ok(*message == deserialized_message)
+}
+
+#[test]
+fn test_serialize_client_hello() {
+    fn property(random: Vec<u8>) -> TestResult {
+        if random.len() > REPLAY_PROTECTION_ARRAY_LENGTH {
+            return TestResult::discard();
+        }
+        let client_hello = ClientHello::new(to_array(&random).unwrap());
+        let result = test_serialize_template(&client_hello);
+        assert!(result.is_ok());
+        TestResult::from_bool(result.unwrap())
+    }
+    quickcheck(property as fn(Vec<u8>) -> TestResult);
```

Comment:
why is the `as` necessary? That seems exactly the same signature

Reply:
Because `quickcheck` requires `Testable` to be implemented.
And it's implemented as follows:
https://github.com/BurntSushi/quickcheck/blob/defde6fb0ce20b0c8c4e672aa9ae821f7d1f5b38/src/tester.rs#L386-L394

Reply:
Added a comment